### PR TITLE
RNA-STAR: additional parameters

### PR DESCRIPTION
**Pull request by jsh58**
More parameters for users to select (output, seed, alignment, and chimeric alignment).

Re: STAR-Fusion preset option -- the recommended parameter '--twopassMode Basic' is not available in STAR 2.4.0, but it is possible to run a second pass by adjusting other parameters.  However, doing so resulted in a failure in test #3, so these parameters have been commented out (lines 81-82).

### CODE REVIEW

**Comment by bgruening** [comment]
@jsh58 thanks for the PR, this looks like a great contribution. 
@yhoogstrate you are using STAR if I'm not misleading right, could you please review this? 

Thanks!

**Comment by yhoogstrate** [comment]
Great work @jsh58, I will take a look at this

**Comment by hrhotz** [comment]
I did some work with 1.5.1 but it all ended up in our Galaxy wrapper for QuasR (a BioC package developed by my boss) and it is useless outside our local setup. I had no time to look at / work on the generic STAR wrapper - sorry.

File: tools/rgrnastar/rg_rnaStar.xml [hunk 1]
```diff
@@ -121,63 +179,138 @@
...
+          </when>
         </conditional>
+        
     </inputs>
 
     <outputs>
         <data format="txt" name="output_log" label="${tool.name} on ${on_string}: log" from_work_dir="Log.final.out"/>
         <data format="interval" name="chimeric_junctions" label="${tool.name} on ${on_string}: starchimjunc" from_work_dir="Chimeric.out.junction">
-            <filter>(params['settingsType'] != 'default' and params['chim_segment_min'] > 0)</filter>
+            <filter>params['settingsType'] == "star_fusion" or ( params['settingsType'] == "full" and params['chim']['chim_select'] == "yes" and params['chim']['segmentMin'] > 0 )</filter>
```
Comment (yhoogstrate) [inline tools/rgrnastar/rg_rnaStar.xml]: `params['settingsType'] in [ "star_fusion" , "full" ]`

File: tools/rgrnastar/rg_rnaStar.xml [hunk 2]
```diff
@@ -198,7 +331,7 @@
         </data>
 
         <data format="bam" name="chimeric_reads" label="${tool.name} on ${on_string}: starmappedchim.bam" from_work_dir="ChimericSorted.bam">
-            <filter>(params['settingsType'] == 'full' and params['chim_segment_min'] > 0)</filter>
+            <filter>params['settingsType'] == "star_fusion" or ( params['settingsType'] == "full" and params['chim']['chim_select'] == "yes" and params['chim']['segmentMin'] > 0 )</filter>
```
Comment (yhoogstrate) [inline tools/rgrnastar/rg_rnaStar.xml]: `params['settingsType'] in [ "star_fusion" , "full" ]`

**Comment by yhoogstrate** [comment]
Hi jsh58,

I have been looking into your changes and I am really impressed by the level of detail you have put in. I like the way how the arguments are now organized within different sections, which reminded me of the section feature. The four conditionals can be replaced by the `<section />` tag:
https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax#A.3Csection.3E_tag_set
If you want to see how they can be implemented, you can take a look at the JBrowse wrapper.

I am referring to these conditionals:
 - Would you like to set output parameters (formatting and filtering)?

And within the extended parameters:
  - Would you like to set seed parameters?
  - Would you like to set alignment parameters?
  - Would you like to set chimeric alignment parameters?

Another new feature is `argument=` in the `<params>` tag set, which would replace
```
<param name="outSAMattributes" label="Extra SAM attributes to include (--outSAMattributes)">
<param name="outSAMstrandField" label="Include strand field flag XS (--outSAMstrandField)">
```

for:

```
<param name="outSAMattributes" label="Extra SAM attributes to include" argument="--outSAMattributes">
<param name="outSAMstrandField" label="Include strand field flag XS" argument="--outSAMstrandField">
```
It would be great to make those changes as well.

Something popped up when I was reading the output section. I saw:
```
<data format="txt" name="output_log" label="${tool.name} on ${on_string}: log" from_work_dir="Log.final.out"/>
```
I suppose a user does not really want to use this file in a pipeline so maybe it's nice to set this to `hidden='True'`? I'm glad to hear your opinion on this. If you like it you could make this change as well.

The tests have passed so further it's ready to be merged.

File: tools/rgrnastar/rg_rnaStar.xml [hunk 3]
```diff
@@ -51,55 +52,112 @@
...
     
     ## Convert aligned reads.
-    samtools view -@ \${GALAXY_SLOTS:-4} -Shb Aligned.out.sam | samtools sort -@ \${GALAXY_SLOTS:-4} - AlignedSorted 2>/dev/null ;
+    && samtools view -@ \${GALAXY_SLOTS:-4} -Shb Aligned.out.sam | samtools sort -@ \${GALAXY_SLOTS:-4} - AlignedSorted 2>/dev/null
     
     ## Convert chimeric reads.
-    #if str($params.settingsType) == "full" and $params.chim_segment_min > 0:
-        samtools view-@ \${GALAXY_SLOTS:-4} -Shb Chimeric.out.sam | samtools sort -@ \${GALAXY_SLOTS:-4} - ChimericSorted 2>/dev/null
+    #if str($params.settingsType) == "star_fusion" or ( str($params.settingsType) == "full" and str($params.chim.chim_select) == "yes" and int($params.chim.segmentMin) > 0 ):
+        ; samtools view -@ \${GALAXY_SLOTS:-4} -Shb Chimeric.out.sam | samtools sort -@ \${GALAXY_SLOTS:-4} - ChimericSorted 2>/dev/null
```
Comment (nsoranzo) [inline tools/rgrnastar/rg_rnaStar.xml]: Please use `&&` to concatenate the commands instead of `;`.

**Comment by nsoranzo** [comment]
In addition to what @yhoogstrate said about the `argument` attribute, the IUC best practices prescribe that "(--argument)" should be added to the `help` attribute, not to the `label`:

https://galaxy-iuc-standards.readthedocs.org/en/latest/best_practices/tool_xml.html#parameter-help

**Comment by jsh58** [comment]
Many comments / responses:

1. Parameter names moved to help messages.  I tried the "argument=" feature, but the name sometimes would not show up.

2. Re: log file, I often use this file to check the counts, so I think it's OK to keep it as an output.

3. Re: <section> instead of <conditional>, I tried this but did not like the results.  The interface wasn't as clear, and the command-line wasn't as clean (all the parameters are specified even if none is changed).

**Comment by yhoogstrate** [comment]
Hi @jsh58 , thanks for the quick reply again.

1. You are right and it seems to be a bug in Galaxy. I've written a small path to solve it so that all `argument=""`-tags become visible again. For the moment, don't change it and lets await that PR.

2. Fair point

3. If you still have the wrapper with the sections, could you make a branch and push them so I can take a look at it? For these particular use cases the goal is to make use of sections at some point in time, so I would like to see what makes it unclear. Maybe also here it might require a change to galaxy?

**Comment by jsh58** [comment]
Hi @yhoogstrate,

I pushed the version with "sections" to the tools-iuc_test branch:
https://github.com/jsh58/tools-iuc/blob/tools-iuc_test/tools/rgrnastar/rg_rnaStar.xml
It is incomplete; the "output" options were not sectionified.

It probably would be improved with different titles, e.g. title="Click to set seed parameters" instead of just title="Seed parameters".  

Also, the command-line is cluttered because we can no longer use 
`#if str( $params.seed.seed_select ) == "yes":
  ## set a bunch of seed-related parameters
#end if`
so, like I said before, all the parameters are specified once "Extended parameter list" is chosen, even if none of the parameters is actually changed.

**Comment by yhoogstrate** [comment]
Hey @jsh58,

Thanks again! I've been looking through it and I get what you mean.
I don't think the part with the if statements is a big deal, since the user has selected to choose advanced settings. So, as you did, removing the if statements like: `if str( $params.chim.chim_select ) == "yes":` is desired. If the default values are identical to those of STAR it shouldn't matter. Anyway, using sections is not part of the guidelines so I would argue that what you have proposed is great and you can leave it as it is.

In the meantime the issue of arguments has been solved (requires commits later than https://github.com/galaxyproject/galaxy/commit/09911084f3b424d4822d4b5baed7fa3ca1a0a3b9). Let me know of you still want to change this? Depending on whether you want to make these changing I think we're ready to merge :). I have added a 4th test case that expands and tests the new conditionals as PR in your channel.

Good work, thanks!

**Comment by jsh58** [comment]
Hi @yhoogstrate,
Re: sections, it is true that the Galaxy output is not affected, because the default values are identical.  But, from a user's perspective, I find it cleaner to have the conditionals to understand how a tool was run.  The command-line is shorter, and the parameters selected make it clear that when chim_select == "yes" and seed_select == "no", the chimera parameters may have been altered from their defaults, but the seed parameters were not.
Re: arguments, I would rather keep the parameters in the help messages at this point.
Ready to merge, from my perspective.  Thanks!

**Comment by yhoogstrate** [comment]
I have updated the tool sheds

**Comment by bgruening** [comment]
Awesome @jsh58 and @yhoogstrate! Thanks a lot!

**Pull request merged**